engineering_tools/carma_script_extensions/swarm

```diff
@@ -0,0 +1,352 @@
+    if [[ -z $carma_base ]]; then
+        __pull_newest_carma_base
+        carma_base=$(__get_most_recent_carma_base)
+    fi
+
+    docker run -it --rm -v carma-config-data:/opt/carma/vehicle/config $carma_base bash
+}
+
+
+swarm-config__status() {
+    if [ ! -z "$(docker volume ls | grep carma-config-data)" ]; then
+        echo "Found carma-config-data volume"
+    else
+        echo "Missing carma-config-data volume, add with \"carma swarm update-config <image tag>\""
+        exit 1
+    fi
+
+    if [ -z "$1" ]; then
+        echo ""
+        docker run --rm -v carma-config-data:/opt/carma/vehicle/config --entrypoint sh busybox:latest -c \
+        'cat /opt/carma/vehicle/config/docker-compose-manager.yml <(echo)
+         cat /opt/carma/vehicle/config/docker-compose-worker.yml <(echo)
+         cat /opt/carma/vehicle/config/docker-compose-background-manager.yml <(echo)
+         cat /opt/carma/vehicle/config/carma.launch <(echo) 
+         cat /opt/carma/vehicle/config/carma.config.js <(echo)
+         cat /opt/carma/vehicle/config/carma.urdf <(echo)'
+        
+    else
+        echo " -- $1:"
+        docker run --rm -v carma-config-data:/opt/carma/vehicle/config --entrypoint sh busybox:latest -c \
+        "cat /opt/carma/vehicle/config/$1"
+    fi
+
+}
+
+
+swarm__attach() {
+    echo "Attaching STDOUT logs to terminal..."
+
+    SWARM_SERVICES=$(docker service ls --format "{{.Name}}")
+
+    # this seems to loop forever and cannot be escaped with ctrl+C
+    # performance will likely be poor due to combined log streams
+    for TASK in $SWARM_SERVICES; do
+        docker service logs --follow --details --tail=10 "$TASK" &
+    done
+
+    sleep infinity
+}
+
+
+swarm__register() {
+    echo "Installing Docker images..."
+
+    if [ "$(docker volume ls | grep carma-registry | sed 's/local     //g')" ]; then
+        echo "Found carma-registry volume"
+    else
+        echo "Missing carma-registry volume, adding now..."
+        docker volume create carma-registry
+    fi
+
+    # if [ "$2" = "-d" ]; then
+    #     local USERNAME=usdotfhwastoldev
+    #     local TAG=$3
+    # elif [ "$2" = "-c" ]; then
+    #     local USERNAME=usdotfhwastolcandidate
+    #     local TAG=$3
+    # else
+    #     local USERNAME=usdotfhwastol
+    #     local TAG=$2
+    # fi
+
+    if [[ -z $1 ]]; then
+        echo "Please specify a tag string to update carma-config-data volume."
+        echo "Done."
+        exit 1
+    fi
+
+    docker service create --constraint 'node.role==manager' -d --volume "carma-registry:/var/lib/registry" -p 5000:5000 --name carma-registry-service registry:2
+    
+    IMAGE_NAMES="$*"
+
+    for SOURCE_IMAGE_NAME in $IMAGE_NAMES
+    do
+    
+    echo "Pushing $SOURCE_IMAGE_NAME to localhost:5000"
+
+    docker tag "$SOURCE_IMAGE_NAME" "localhost:5000/$SOURCE_IMAGE_NAME"
+    docker push "localhost:5000/$SOURCE_IMAGE_NAME"
+    docker rmi "localhost:5000/$SOURCE_IMAGE_NAME"
+
+    done
+}
+
+
+swarm__stop() {
```

Comment:
Is there a difference between stopping a swarm and tearing it down? If so, should we have commands to let users do that?

Reply:
For our purposes I don't think so. Technically Swarm can remain initialized but without a stack deployed to it there's no functionality/need for it to be running in the background.